provider/bidengine/pricing.go

```diff
@@ -60,9 +172,175 @@ func calculatePriceRange(gspec *dtypes.GroupSpec) (sdk.Coin, sdk.Coin) {
 		cmax = rmax.Amount
 	}
 
+	if cmin.IsZero() {
+		cmin = sdk.NewInt(1)
+	}
+
 	if cmax.IsZero() {
 		cmax = sdk.NewInt(1)
 	}
 
 	return sdk.NewCoin(rmax.Denom, cmin), sdk.NewCoin(rmax.Denom, cmax)
 }
+
+type shellScriptPricing struct {
+	path         string
+	processLimit chan int
+	runtimeLimit uint
+}
+
+var errPathEmpty = errors.New("script path cannot be the empty string")
+var errProcessLimitZero = errors.New("process limit must be greater than zero")
+var errProcessRuntimeLimitZero = errors.New("process runtime limit must be greater than zero")
+
+func MakeShellScriptPricing(path string, processLimit uint, runtimeLimit uint) (BidPricingStrategy, error) {
+	if len(path) == 0 {
+		return nil, errPathEmpty
+	}
+	if processLimit == 0 {
+		return nil, errProcessLimitZero
+	}
+	if runtimeLimit == 0 {
+		return nil, errProcessRuntimeLimitZero
+	}
+
+	result := shellScriptPricing{
+		path:         path,
+		processLimit: make(chan int, processLimit),
+		runtimeLimit: runtimeLimit,
+	}
+
+	// Use the channel as a semaphore to limit the number of processes created for computing bid processes
+	// Most platforms put a limit on the number of processes a user can open. Even if the limit is high
+	// it isn't a good idea to open thuosands of processes.
+	for i := uint(0); i != processLimit; i++ {
+		result.processLimit <- 0
+	}
+
+	return result, nil
+}
+
+type dataForScriptElement struct {
+	Memory  uint64 `json:"memory"`
+	CPU     uint64 `json:"cpu"`
+	Storage uint64 `json:"storage"`
+	Count   uint32 `json:"count"`
+}
+
+var errCalculatePriceCancelled = errors.New("calculate price cancelled")
+var errCalculatePriceTimeout = errors.New("calculate price timed out")
+
+func (ssp shellScriptPricing) calculatePrice(gspec *dtypes.GroupSpec, lc lifecycle.Lifecycle) (sdk.Coin, error) {
```

Comment:
this should take a `context.Context`.  Then you can layer on execution timeout with [`context.WithTimeout`](https://golang.org/pkg/context/#WithTimeout) and halt the program with [`CommandContext()`](https://golang.org/pkg/os/exec/#CommandContext).

Reply:
If I understand what you're describing, I'd have to create a context in `provider/bidengine/order.go` and then pass it into this function.  Then when the `provider/bidengine/order.go` hits the case of `<-o.lc.ShutdownRequest()`  I'd have cancel the context _if it exists at that time_. 

I'm not sure how I understand this is simpler.

Reply:
Just make a context at the beginning of run and always cancel it before returning from run. 

Context is the common lifetime object to pass around.  Using it plus runcontext makes it so that you don't need to do all this grouting handling in here.

It's also just not a good idea to leak the lifecycle object as you've done.  There is a reader interface you can use but please just pass around context instead.

Reply:
This change has been made, the `order` has its own context and each pricing operation that calls a script makes a child context with a deadline to handle timeouts.